### DISCUSSION TITLE
How to do a basic config/parameter validation

### DISCUSSION THREAD

**Discussion by hogru** [facebookresearch/hydra#2259] [Q&A] [answered]
I found a couple of ways to validate my configuration parameters (`Enum`s, custom `OmegaConf` resolvers, `OmegaConf.to_object())`. However I fail to do even a simple type check. I would expect this to raise an exception or at least give me type hints in the IDE, but apparently I again don't see the forest for the trees. Can you help me in telling me what's wrong with this code?

```python
from dataclasses import dataclass

import hydra
from omegaconf import OmegaConf


@dataclass
class MyParam:
    my_param: int  # I want my_param to be an int


@hydra.main(config_path="./", config_name="test", version_base=None)
def main(cfg: MyParam) -> None:
    print(OmegaConf.to_yaml(cfg))  # my_param: this_is_a_string
    print(type(cfg.my_param))      # <class 'str'>


if __name__ == "__main__":
    main()
```

And this is `test.yaml`

```yaml
my_param: this_is_a_string
```

**Comment by Jasha10** [comment, marked as answer]
See the docs on [Structured Config schema](https://hydra.cc/docs/tutorials/structured_config/schema/).
You'll need to use a schema if you want to take advantage of OmegaConf's runtime type checking.

**Reply by hogru** [reply]
I thought that I had a schema… until it did not catch type errors. The consequence of this problem was my attempt to provide a minimum “working” example above and “loosing” the schema during the process. Your hint encouraged me to take another look at the structured config schema, and I realized that I had missed the  _correct_ ` cs.store()` calls along with the corresponding `defaults` entries. I must admit this is one of the more complex frameworks I came across, I think I am only starting to fully understand what's going on behind the scenes (schema, multiple .yaml and .py files, …). But now I get `ValidationError` exceptions, which is what I wanted. Thank you! Next: \_\_post\_init\_\_() ;-)

**Reply by Jasha10** [reply]
> I must admit this is one of the more complex frameworks I came across

Hydra is a framework for building frameworks ;)

Note you'll need to use [`OmegaConf.to_object`](https://omegaconf.readthedocs.io/en/2.1_branch/usage.html#omegaconf-to-object) if you want `MyParam.__post_init__` to take effect. `OmegaConf.to_object` will convert your config from an instance of `DictConfig` to an instance of `MyParams`. For example:

```python
...

@dataclass
class MyParam:
    ...

    def __post_init__(self):
        ...

cs.store(name="my_param_schema", node=MyParam)

@hydra.main(...)
def main(_cfg: DictConfig):
    assert isinstance(_cfg, DictConfig)
    cfg: MyParams = OmegaConf.to_object(_cfg)  # __post_init__ runs here
    ...
    
```

**Reply by hogru** [reply]
Thanks, @Jasha10, your “meta comment” made me feel a bit more capable 😉 

You already "told" me about `OmegaConf.to_object()` in this [discussion](https://github.com/facebookresearch/hydra/issues/981#issuecomment-1007307761). This works nicely and I use it to do some validations I can't (or don't know how to) do within the `dataclass`es, e.g. `assert sum(random_split) = 1.`. But for actually accessing the config values, I keep the structured config because I feel its syntax is nicer to use. So, the only thing that remains is that I would like to convert some config values to e.g., `NamedTuples`. This is not strictly necessary, but just easier to read. I do this in `main()` now because doing it in `__post_init__()` works on the data structure that I don't use to access the config values (for the reason above). Thanks again for your help 👍
